Framework/API/inc/MantidAPI/Algorithm.tcc

```diff
@@ -0,0 +1,133 @@
+      dynamic_cast<WorkspaceProperty<T> *>(getPointerToProperty(propertyName));
+  auto *indexTypeProp = dynamic_cast<IndexTypeProperty *>(
+      getPointerToProperty(propertyName + "IndexType"));
+
+  declareProperty(Kernel::make_unique<IndexProperty>(propertyName + "IndexSet",
+                                                     *wsProp, *indexTypeProp));
+
+  m_reservedList.push_back(propertyName);
+  m_reservedList.push_back(propertyName + "IndexType");
+  m_reservedList.push_back(propertyName + "IndexSet");
+}
+
+/** Mechanism for setting the index property with a workspace shared pointer.
+* This method can only be used if T1 is convertible to a MatrixWorkspace and
+* T2 is either std::string or std::vector<int>
+
+@param name Property name
+@param wksp Workspace as a pointer
+@param type Index type IndexType::WorkspaceIndex or IndexType::SpectrumNum
+@param list List of indices to be used.
+*/
+template <typename T1, typename T2, typename, typename>
+void Algorithm::setIndexProperty(const std::string &name,
+                                 const boost::shared_ptr<T1> &wksp,
+                                 IndexType type, const T2 &list) {
+  if (!isCompoundProperty(name))
+    throw std::runtime_error("Algorithm::setIndexProperty can only be used "
+                             "with properties declared using "
+                             "declareIndexProperty.");
+  auto *wsProp =
+      dynamic_cast<WorkspaceProperty<T1> *>(getPointerToProperty(name));
+  auto *indexTypeProp = dynamic_cast<IndexTypeProperty *>(
+      getPointerToProperty(name + "IndexType"));
+  auto *indexProp =
+      dynamic_cast<IndexProperty *>(getPointerToProperty(name + "IndexSet"));
+
+  *wsProp = wksp;
+
+  *indexTypeProp = type;
+
+  *indexProp = list;
+}
+
+/** Mechanism for setting the index property with a workspace shared pointer.
+* This method can only be used if T1 is convertible to a MatrixWorkspace and
+* T2 is either std::string or std::vector<int>
+
+@param name Property name
+@param wsName Workspace as a pointer
+@param type Index type IndexType::WorkspaceIndex or IndexType::SpectrumNum
+@param list List of indices to be used.
+*/
+template <typename T1, typename T2, typename, typename>
+void Algorithm::setIndexProperty(const std::string &name,
+                                 const std::string &wsName, IndexType type,
+                                 const T2 &list) {
+  if (!isCompoundProperty(name))
+    throw std::runtime_error("Algorithm::setIndexProperty can only be used "
+                             "with properties declared using "
+                             "declareIndexProperty.");
+  auto *wsProp =
+      dynamic_cast<WorkspaceProperty<T1> *>(getPointerToProperty(name));
+  auto *indexTypeProp = dynamic_cast<IndexTypeProperty *>(
+      getPointerToProperty(name + "IndexType"));
+  auto *indexProp =
+      dynamic_cast<IndexProperty *>(getPointerToProperty(name + "IndexSet"));
+
+  wsProp->setValue(wsName);
```

Comment:
Can't you just put this single line in a templated function to distinguish betwen the `std::string` and `shared_ptr` cases? Instead of duplicating the whole big method.